c10/core/Layout.h

```diff
@@ -6,11 +6,12 @@
 #include <iostream>
 
 namespace c10 {
-enum class Layout : int8_t { Strided, Sparse, Mkldnn };
+enum class Layout : int8_t { Strided, Sparse, Mkldnn, Texture4C };
```

Comment:
Is Vulkan going to support multiple layouts? If not - I'd suggest to just use 'Strided' (aka Dense) which is kind of default. Our layouts are sadly not super clean, but it's less modifications this way

Reply:
In this diff we have only one opaque texture layout.
The next step will be to add Strided buffer layout.

I think we will expose 'Strided' buffer layout for users, to write custom ops, to support all the logic etc.

I do not know if we want to expose Texture4c, when we have 'Strided', probably that will be our internal implementation details.  Anyway we will need conversion strided buffer -> texture to not write stride-iteration logic inside shaders for ops. But maybe @dreiss , @AshkanAliabadi  have different opinion about it.

Reply:
Ah, interesting. Do we want to allow users to explicitly convert to/from Texture4c? Can there be a kernel that takes arguments in different format (one in Texture4c and one in strided)?

Basically we need to decide which option of leveraging current tensor primitives makes most sense:
1) make Texture4c choice completely opaque, e.g. some operators return textures instead of buffers but it's hidden in TensorImpl details. So that if I call contiguous or anything else the texture gets silently converted to buffer
2) make Texture4c the layout as you do here. In this case it'd be user-visible and the user can manipulate conversion back and forth explicitly
3) make Texture4c separate device. It's almost the same as layout, but we can leverage dispatcher to model mixed kernels and new textures allocation would go a bit of a different path (with its own allocator). You'd be also able to do `.to('vulcan-texture')` instead of a separate method

Above is an important decision to make - feel free to ping me offline too for a detailed discussion